File: packages/tonic-core/src/weaveGit/verify.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { currentLines } from "../core";
import { deserializeState, serializeState } from "../state";
import type { TonicGitManifest } from "./types";
import { normalizeLf, sha256HexBytes, sha256HexUtf8 } from "./hashutil";
import { parseManifestJson } from "./manifest";
import { gitStagedPaths } from "./gitStaging";
import { findLfsPointersUnder } from "./lfs";

export const DEFAULT_WEAVE_ROOT = path.join(".tonic", "weave");

export type VerifyCheck = { id: string; ok: boolean; message: string; path?: string };
export type VerifyError = { code: string; message: string; path?: string };

export type VerifyResult = {
  ok: boolean;
  status: string;
  strict: boolean;
  errors: VerifyError[];
  checks: VerifyCheck[];
};

/** Match ``VerifyResult.report_dict`` status logic in Python ``verify.py``. */
export function verifyReportStatus(ok: boolean, strict: boolean, errorCount: number): "ok" | "failed" | "degraded" {
  if (ok) {
    return "ok";
  }
  if (strict || errorCount === 0) {
    return "failed";
  }
  return "degraded";
}

export function verifyReportDict(repoRoot: string, vr: VerifyResult): Record<string, unknown> {
  return {
    schema: "tonic-weave-verify-report",
    version: "1",
    status: verifyReportStatus(vr.ok, vr.strict, vr.errors.length),
    strict: vr.strict,
    repo_root: path.resolve(repoRoot),
    checks: vr.checks,
    errors: vr.errors,
  };
}

export function weaveBlobPath(weaveRoot: string, weaveSerializedSha: string): string {
  return path.join(weaveRoot, "blobs", weaveSerializedSha);
}

export function verifyPathLocal(
  repoRoot: string,
  relPath: string,
  entry: Record<string, unknown>,
  weaveRoot: string,
): string[] {
  const errs: string[] = [];
  const p = path.join(repoRoot, relPath);
  if (!fs.existsSync(p) || !fs.statSync(p).isFile()) {
    errs.push(`${relPath}: missing file`);
    return errs;
  }
  const raw = fs.readFileSync(p, "utf8");
  const norm = normalizeLf(raw);
  const lines = norm.length ? norm.split("\n") : [];
  if (lines.length && lines[lines.length - 1] === "") {
    lines.pop();
  }
  const textSha = sha256HexUtf8(norm);
  const wantText = entry.text_blob_sha;
  if (typeof wantText !== "string" || textSha !== wantText) {
    errs.push(`${relPath}: text_blob_sha mismatch (got ${textSha})`);
  }
  const wss = entry.weave_serialized_sha;
  if (typeof wss !== "string") {
    errs.push(`${relPath}: invalid weave_serialized_sha in manifest`);
    return errs;
  }
  const blob = weaveBlobPath(weaveRoot, wss);
  if (!fs.existsSync(blob)) {
    errs.push(`${relPath}: missing weave blob at ${blob}`);
    return errs;
  }
  const weaveBytes = fs.readFileSync(blob);
  if (sha256HexBytes(weaveBytes) !== wss) {
    errs.push(`${relPath}: weave blob hash mismatch`);
    return errs;
  }
  let stateS: string;
  try {
    stateS = weaveBytes.toString("utf8");
    deserializeState(stateS);
  } catch (e) {
    errs.push(`${relPath}: deserialize_state failed: ${e}`);
    return errs;
  }
  const roundtrip = Buffer.from(serializeState(deserializeState(stateS)), "utf8");
  if (sha256HexBytes(roundtrip) !== wss) {
    errs.push(`${relPath}: weave round-trip sha mismatch`);
  }
  let cl: string[];
  try {
    cl = currentLines(stateS);
  } catch (e) {
    errs.push(`${relPath}: current_lines failed: ${e}`);
    return errs;
  }
  if (cl.join("\n") !== lines.join("\n")) {
    errs.push(`${relPath}: working tree text does not match weave current_lines`);
  }
  return errs;
}

export function verifyManifest(
  repoRoot: string,
  manifest: TonicGitManifest,
  options: { weaveRoot?: string; onlyPaths?: Set<string> | null; strict?: boolean } = {},
): VerifyResult {
  const wr = options.weaveRoot ?? path.join(repoRoot, DEFAULT_WEAVE_ROOT);
  const checks: VerifyCheck[] = [];
  const errors: VerifyError[] = [];
  const strict = options.strict ?? false;
  const only = options.onlyPaths ?? null;
  const paths = manifest.paths;
  for (const [rel, ent] of Object.entries(paths)) {
    if (only !== null && !only.has(rel)) {
      continue;
    }
    if (typeof ent !== "object" || ent === null) {
      errors.push({ code: "bad_entry", message: `${rel}: not an object`, path: rel });
      continue;
    }
    const pe = verifyPathLocal(repoRoot, rel, ent as Record<string, unknown>, wr);
    const ok = pe.length === 0;
    checks.push({ id: `path:${rel}`, ok, message: pe.join("; ") || "", path: rel });
    for (const m of pe) {
      errors.push({ code: "verify_path", message: m, path: rel });
    }
  }
  if (strict) {
    for (const [rel, ent] of Object.entries(paths)) {
      if (only !== null && !only.has(rel)) {
        continue;
      }
      if (typeof ent === "object" && ent !== null && (ent as { degraded?: boolean }).degraded === true) {
        const msg = `${rel}: degraded manifest row not allowed in strict mode`;
        errors.push({ code: "strict_degraded", message: msg, path: rel });
        checks.push({ id: `strict:${rel}`, ok: false, message: msg, path: rel });
      }
    }
  }
  const ok = errors.length === 0;
  const status = verifyReportStatus(ok, strict, errors.length);
  return { ok, status, strict, errors, checks };
}

/** Append LFS pointer errors like Python ``cmd_weave_verify`` when strict or ``--check-lfs``. */
export function applyLfsPointerChecks(
  vr: VerifyResult,
  repoRoot: string,
  opts: { checkLfs: boolean },
): VerifyResult {
  if (!opts.checkLfs && !vr.strict) {
    return vr;
  }
  const objs = path.join(repoRoot, ".tonic", "objects");
  const pointers = findLfsPointersUnder(objs);
  if (!pointers.length) {
    return vr;
  }
  const errors = [...vr.errors];
  const checks = [...vr.checks];
  for (const p of pointers) {
    errors.push({ code: "lfs_pointer", message: `LFS pointer not smudged: ${p}`, path: p });
    checks.push({ id: `lfs:${p}`, ok: false, message: "pointer present", path: p });
  }
  const ok = false;
  const status = verifyReportStatus(ok, vr.strict, errors.length);
  return { ok, status, strict: vr.strict, errors, checks };
}

export function verifyStaged(
  repoRoot: string,
  manifestJson: string,
  options: { weaveRoot?: string; strict?: boolean } = {},
): VerifyResult {
  const manifest = parseManifestJson(manifestJson);
  const staged = gitStagedPaths(repoRoot);
  let onlyPaths: Set<string> | null = null;
  if (staged.size === 0) {
    onlyPaths = null;
  } else {
    const manifestPaths = new Set(Object.keys(manifest.paths));
    const inter = new Set([...manifestPaths].filter((k) => staged.has(k)));
    if (inter.size === 0) {
      const strict = options.strict ?? false;
      return {
        ok: true,
        status: "ok",
        strict,
        checks: [{ id: "staged", ok: true, message: "no staged tonic paths in manifest" }],
        errors: [],
      };
    }
    onlyPaths = inter;
  }
  return verifyManifest(repoRoot, manifest, { ...options, onlyPaths });
}

export function verifyReportJson(repoRoot: string, manifestJson: string, opts?: object): string {
  const m = parseManifestJson(manifestJson);
  const vr = verifyManifest(repoRoot, m, opts as { weaveRoot?: string; onlyPaths?: Set<string> | null; strict?: boolean });
  return JSON.stringify(verifyReportDict(repoRoot, vr), null, 2);
}

File: merge-tonic-lib/tonic/weave_git/verify.py
from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass, field
from pathlib import Path

from tonic.core import current_lines
from tonic.core.state import deserialize_state, serialize_state

from .hashutil import canonical_text_lines, normalize_lf, sha256_hex_bytes, sha256_hex_utf8
from .manifest import load_manifest_path, parse_manifest_json
from .types import TonicGitManifest


DEFAULT_WEAVE_ROOT = Path(".tonic/weave")


@dataclass
class VerifyResult:
    ok: bool
    status: str
    strict: bool
    errors: list[dict[str, str]] = field(default_factory=list)
    checks: list[dict[str, object]] = field(default_factory=list)

    def report_dict(self, repo_root: str) -> dict[str, object]:
        return {
            "schema": "tonic-weave-verify-report",
            "version": "1",
            "status": "ok" if self.ok else ("failed" if self.strict or not self.errors else "degraded"),
            "strict": self.strict,
            "repo_root": repo_root,
            "checks": self.checks,
            "errors": self.errors,
        }


def _git_staged_paths(repo: Path) -> set[str]:
    try:
        out = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "-z"],
            cwd=repo,
            capture_output=True,
            text=True,
            check=False,
        )
    except OSError:
        return set()
    if out.returncode != 0:
        return set()
    raw = out.stdout
    if not raw:
        return set()
    return {p.replace("\\", "/") for p in raw.split("\0") if p}


def weave_blob_path(weave_root: Path, weave_serialized_sha: str) -> Path:
    return weave_root / "blobs" / weave_serialized_sha


def verify_path_local(
    repo_root: Path,
    rel_path: str,
    entry: dict[str, object],
    weave_root: Path,
) -> list[str]:
    errs: list[str] = []
    p = repo_root / rel_path
    if not p.is_file():
        errs.append(f"{rel_path}: missing file")
        return errs
    raw = p.read_text(encoding="utf-8")
    norm = normalize_lf(raw)
    text_sha = sha256_hex_utf8(norm)
    want_text = entry.get("text_blob_sha")
    if not isinstance(want_text, str) or text_sha != want_text:
        errs.append(f"{rel_path}: text_blob_sha mismatch (got {text_sha})")

    wss = entry.get("weave_serialized_sha")
    if not isinstance(wss, str):
        errs.append(f"{rel_path}: invalid weave_serialized_sha in manifest")
        return errs
    blob = weave_blob_path(weave_root, wss)
    if not blob.is_file():
        errs.append(
            f"{rel_path}: missing weave blob at {blob}; "
            f"weave blobs may not be in git — run `hf weave prefetch` / `hf weave push` or commit `.tonic/weave`"
        )
        return errs
    weave_bytes = blob.read_bytes()
    if sha256_hex_bytes(weave_bytes) != wss:
        errs.append(f"{rel_path}: weave blob hash mismatch")
        return errs
    try:
        state_s = weave_bytes.decode("utf-8")
        state = deserialize_state(state_s)
    except Exception as e:  # noqa: BLE001 — surface to caller
        errs.append(f"{rel_path}: deserialize_state failed: {e}")
        return errs
    roundtrip = serialize_state(state).encode("utf-8")
    if sha256_hex_bytes(roundtrip) != wss:
        errs.append(f"{rel_path}: weave round-trip sha mismatch")
    lines = canonical_text_lines(raw)
    try:
        cl = current_lines(state_s)
    except Exception as e:  # noqa: BLE001
        errs.append(f"{rel_path}: current_lines failed: {e}")
        return errs
    if cl != lines:
        errs.append(f"{rel_path}: working tree text does not match weave current_lines")
    return errs


def verify_manifest(
    repo_root: Path,
    manifest: TonicGitManifest,
    *,
    weave_root: Path | None = None,
    only_paths: set[str] | None = None,
    strict: bool = False,
) -> VerifyResult:
    wr = weave_root or (repo_root / DEFAULT_WEAVE_ROOT)
    checks: list[dict[str, object]] = []
    errors: list[dict[str, str]] = []
    paths = manifest.get("paths", {})
    for rel, entry in paths.items():
        if only_paths is not None and rel not in only_paths:
            continue
        if not isinstance(entry, dict):
            errors.append({"code": "bad_entry", "message": f"{rel}: not an object", "path": rel})
            continue
        pe = verify_path_local(repo_root, rel, entry, wr)
        ok = len(pe) == 0
        checks.append({"id": f"path:{rel}", "ok": ok, "message": "; ".join(pe) if pe else "", "path": rel})
        for m in pe:
            errors.append({"code": "verify_path", "message": m, "path": rel})
    # strict: degraded entries forbidden
    if strict:
        for rel, entry in paths.items():
            if only_paths is not None and rel not in only_paths:
                continue
            if isinstance(entry, dict) and entry.get("degraded") is True:
                msg = f"{rel}: degraded manifest row not allowed in strict mode"
                errors.append({"code": "strict_degraded", "message": msg, "path": rel})
                checks.append({"id": f"strict:{rel}", "ok": False, "message": msg, "path": rel})
    ok = len(errors) == 0
    status = "ok" if ok else "failed"
    return VerifyResult(ok=ok, status=status, strict=strict, errors=errors, checks=checks)


def verify_staged(
    repo_root: Path,
    manifest_path: Path,
    *,
    weave_root: Path | None = None,
    strict: bool = False,
) -> VerifyResult:
    manifest = load_manifest_path(manifest_path)
    staged = _git_staged_paths(repo_root)
    if not staged:
        # No git staged files: verify all manifest paths (local dev / tests)
        only = None
    else:
        manifest_paths = set(manifest.get("paths", {}).keys())
        only = manifest_paths & staged
        if not only:
            return VerifyResult(
                ok=True,
                status="ok",
                strict=strict,
                checks=[{"id": "staged", "ok": True, "message": "no staged tonic paths in manifest"}],
            )
    return verify_manifest(repo_root, manifest, weave_root=weave_root, only_paths=only, strict=strict)


def verify_report_json(repo_root: Path, manifest_json: str, **kwargs: object) -> str:
    m = parse_manifest_json(manifest_json)
    vr = verify_manifest(repo_root, m, **kwargs)  # type: ignore[arg-type]
    return json.dumps(vr.report_dict(str(repo_root.resolve())), indent=2)
